add more logic to MatrixUnionRows

catoverdrive opened the pull request:
basically, it now looks at the partitioner for the MatrixValues of its children when it executes, and only shuffles if rvds are non-disjoint. It removes empty rvds and separates children into non-disjoint piles, shuffles those independently, and then concatenates them together in sorted order.

(I'm not sure how this behaves relative to the old behavior (coerce everything together) when we have multiple shuffles that need to happen, but at least filtering out empty RVDs helps a lot in e.g. the split-multi case where none of the variants need to be moved.)

I think the (a?) better optimization would be to add smarter partitioning so that we're adjusting partition bounds and not really shuffling under most circumstances, but I thought I'd PR this first since at least there's a warning in python about unioning non-disjoint datasets (where this will avoid the scan).

I also added tests for MatrixUnionRows and pulled out one of the simplify rules; TableUnion does an unsorted union (I'm not sure this is actually correct behavior now if the tables have keys) and pulling the union outside of a MatrixRowsTable can cause the ordering to be wrong.

catoverdrive commented on src/test/scala/is/hail/expr/ir/MatrixIRSuite.scala:
```diff
@@ -62,19 +69,52 @@ class MatrixIRSuite extends SparkSuite {
+
+  @DataProvider(name="unionRowsData")
+  def unionRowsData(): Array[Array[Any]] = Array(
+    Array(FastIndexedSeq(0 -> 1, 5 -> 7)),
```
yeah :( Do we have a good way of constructing empty MatrixTables via IR (or even just in general)? I was trying to figure out how to do this but I'm not sure I know how.

patrick-schultz replied:
Good question. My first thought is to manually make a `MatrixValue`, and maybe package that up in a `MatrixValue.empty` method.

catoverdrive commented:
@patrick-schultz I've updated this PR so that OrderedRVD.union() uses better logic to avoid the shuffling the RVDs that it's trying to merge. I did this by merging the partitioners so that each resulting partition only gets information from one original partition per input RVD.

I've removed the unionDisjoint stuff because this merge logic should automatically do that in the case of disjoint RVDs.

patrick-schultz commented on src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala:
```diff
@@ -195,4 +195,65 @@ object OrderedRVDPartitioner {
+  // takes range bounds from n partitioners and splits them such that each
+  // resulting partition only comes from one partition per original partitioner.
+  def mergePartitioners(p1: OrderedRVDPartitioner, p2: OrderedRVDPartitioner): OrderedRVDPartitioner = {
+    require(p1.pkType == p2.pkType)
```
I think that this assertion should really be comparing keys, not partition keys. See #4040 for a bug caused by a similar assertion. (It's confusing right now, but I'm working to clean things up.)

patrick-schultz commented on src/main/scala/is/hail/expr/ir/MatrixIR.scala:
```diff
@@ -1979,7 +1979,10 @@ case class MatrixUnionRows(children: IndexedSeq[MatrixIR]) extends MatrixIR {
-    values.head.copy(rvd = OrderedRVD.union(values.map(_.rvd)))
+    val rvds = values.map(  _.rvd)
+    val first = rvds.head
+    require(rvds.tail.forall(_.partitioner.pkType == first.partitioner.pkType))
```
This should also be comparing keys, not partition keys.
